File: LeetCode/DataStruct/JS/Leetcode953-verifyingAnAlienDictionary.js
var isAlienSorted = function (words, order) {
    let n = order.length;
    let mp = new Map();
    for (let i = 0; i < n; i++) mp.set(order[i], i);
    for (let i = 1; i < words.length; i++) {
        const l = words[i - 1], r = words[i];
        let [x, y] = [0, 0];
        while (x < l.length && y < r.length) {
            if (mp.get(l[x]) > mp.get(r[y])) return false;
            if (mp.get(l[x]) < mp.get(r[y])) break;
            x++; y++;
        }
        if (x < l.length && y == r.length) return false;
    }
    return true;
};

File: LeetCode/DataStruct/Golang/Leetcode953-verifyingAnAlienDictionary.go
package sbgo

func isAlienSorted(words []string, order string) bool {
	n, mp := len(order), map[byte]int{}
	for i := 0; i < n; i++ {
		mp[order[i]] = i
	}
	for i := 1; i < len(words); i++ {
		l, r := words[i-1], words[i]
		x, y := 0, 0
		for x < len(l) && y < len(r) {
			if mp[l[x]] > mp[r[y]] {
				return false
			}
			if mp[l[x]] < mp[r[y]] {
				break
			}
			x++
			y++
		}
		if x < len(l) && y == len(r) {
			return false
		}
	}
	return true
}
